### DISCUSSION TITLE
Undefined.SYNO.Core - creating new functions

### DISCUSSION THREAD

**Discussion by oniek** [N4S4/synology-api#163] [Request]
Hi!

I am doing some simple script for reporting permissions and I came across your ready-made API libs for Syno

Unfortunately it does work - I found libs and methods within /usr/syno/synoman/webapi for SYNO.Core.Share.Permission (list method) and SYNO.Core.AppPriv 
then I have added them into core_sys_info.py (e.g. as copy of shared_folders_info) and I am getting:
- Undefined.SYNO.Core.Share.Permission.Error
- Undefined.SYNO.Core.AppPriv.Error
when creating functions that work did you also face such an error that you somehow fixed?
Or is it just not finished by Synology?

**Comment by N4S4** [comment]
Hello,

Could you show me the function you tried to add? So I can better see what's happening

**Comment by oniek** [comment]
I guess there is missing some additional arguments and maybe wrong used :/
- I simply wanted to get application permissions per users/groups like in /etc/synoappprivilege.db
- and share folders permissions per users/groups (like synoacltool -get "/volume1/${directory}")

A) Share folder permission check

- synology_api/core_sys_info.py
```
    def shared_folders_permission(self) -> dict[str, object] | str:
        api_name = 'SYNO.Core.Share.Permission'
        info = self.core_list[api_name]
        api_path = info['path']
        req_param = {'version': info['maxVersion'], 'method': 'list'}

        return self.request_data(api_name, api_path, req_param)
```


- /usr/syno/synoman/webapi# cat SYNO.Core.Share.lib
```
...
"SYNO.Core.Share.Permission": {"allowUser": ["admin.local", "admin.domain", "admin.ldap"], "appPriv": "", "authLevel": 1, "lib": "lib/SYNO.Core.Share.so", "maxVersion": 1, "methods": {"1": [{"list": {"actionPrivileges": ["shared_folder"], "allowDemo": true, "grantByDefault": true}}, {"list_by_user": {"actionPrivileges": ["shared_folder"], "allowDemo": true, "grantByDefault": true}}, {"list_by_group": {"actionPrivileges": ["shared_folder"], "allowDemo": true, "grantByDefault": true}}, {"set": {"actionPrivileges": ["shared_folder"], "allowDemo": true, "grantByDefault": true}}, {"set_by_user_group": {"actionPrivileges": ["shared_folder"], "allowDemo": true, "grantByDefault": true}}]}, "minVersion": 1, "priority": -10, "priorityAdj": 0}
...
```

B) Application permissions

- synology_api/core_sys_info.py
```
    def application_priv(self) -> dict[str, object] | str:
        api_name = 'SYNO.Core.AppPriv'
        info = self.core_list[api_name]
        api_path = info['path']
        req_param = {'version': info['maxVersion'], 'method': 'list'}

        return self.request_data(api_name, api_path, req_param)
```
 
- /usr/syno/synoman/webapi# cat SYNO.Core.AppPriv.lib
```
...
"SYNO.Core.AppPriv.App": {"allowUser": ["admin.local", "admin.ldap", "admin.domain"], "appPriv": "", "authLevel": 1, "lib": "lib/SYNO.Core.AppPriv.so", "maxVersion": 3, "methods": {"1": [], "2": [{"preview": {"actionPrivileges": ["service"], "allowDemo": true, "grantByDefault": true}}, {"allowed": {"allowDemo": true, "grantByDefault": true}}, {"list": {"actionPrivileges": ["service"], "allowDemo": true, "grantByDefault": true}}], "3": [{"list": {"actionPrivileges": ["service"], "allowDemo": true}}]}, "minVersion": 1, "priority": -10, "priorityAdj": 0},
...
```

**Comment by N4S4** [comment]
well For sure there is something missing, I cannot say what, unfortunately there is not many docs. Have you tried to use the browser dev tool to sniff the request while checking user apps permissions? I'll move this topic to requests

**Comment by N4S4** [comment]
closed as already implemented
